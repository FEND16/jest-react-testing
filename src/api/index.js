export const URL = `https://api.fixer.io`;

/**
 * Loads the currency rates from the API and 
 * returns it in json-format as a promise
 * @param {String} date from which date
 * @param {String} base in what currency
 * @returns {Promise}
 */
export function loadRates(date = 'latest', base = "EUR") {
  return fetch(`${URL}/${date}?base=${base}`).then(data => data.json());
};

/**
 * Takes an object and returns an array of
 * objects with the structure: { key: '', value: ''}
 * @param {Object} obj object to convert
 * @returns {Array} 
 */
export function mapObjectToArray(rates){
  if(typeof rates !== 'object'){
    throw new Error("Not an object");
  }
  return Object.entries(rates)
    .map(entry => Object.assign({ value: entry[1]}, { key: entry[0] }));
}

export function sortArrayByName(arrayToSort){
  return arrayToSort.sort((a,b) => a.key > b.key);
}

export function sortArrayByValue(arrayToSort){
  return arrayToSort.sort((a,b) => a.value > b.value);
}