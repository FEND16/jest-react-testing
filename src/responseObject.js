export default {
  "base": "EUR",
  "date": "2017-12-27",
  "rates": {
    "AUD": 1.5316,
    "BGN": 1.9558,
    "BRL": 3.9196,
    "CAD": 1.5028,
    "CHF": 1.1772,
    "CNY": 7.7975,
    "CZK": 25.84,
    "DKK": 7.445,
    "GBP": 0.88593,
    "HKD": 9.2953,
    "HRK": 7.5398,
    "HUF": 310.99,
    "IDR": 16123.0,
    "ILS": 4.1432,
    "INR": 76.312,
    "JPY": 134.7,
    "KRW": 1277.7,
    "MXN": 23.522,
    "MYR": 4.8481,
    "NOK": 9.8605,
    "NZD": 1.6825,
    "PHP": 59.43,
    "PLN": 4.1884,
    "RON": 4.6499,
    "RUB": 68.597,
    "SEK": 9.8727,
    "SGD": 1.5937,
    "THB": 39.027,
    "TRY": 4.5475,
    "USD": 1.1895,
    "ZAR": 14.811
  }
}