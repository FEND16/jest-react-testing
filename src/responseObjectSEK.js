export default {
  "base": "SEK",
  "date": "2007-02-27",
  "rates": {
    "AUD": 0.18056,
    "BGN": 0.21153,
    "CAD": 0.16612,
    "CHF": 0.17503,
    "CNY": 1.1077,
    "CYP": 0.062665,
    "CZK": 3.0784,
    "DKK": 0.80612,
    "EEK": 1.6923,
    "GBP": 0.072864,
    "HKD": 1.1181,
    "HRK": 0.79496,
    "HUF": 27.539,
    "IDR": 1301.0,
    "ISK": 9.4441,
    "JPY": 17.054,
    "KRW": 134.35,
    "LTL": 0.37344,
    "LVL": 0.076184,
    "MTL": 0.046431,
    "MYR": 0.50016,
    "NOK": 0.87162,
    "NZD": 0.20263,
    "PHP": 6.8969,
    "PLN": 0.42319,
    "RON": 0.36645,
    "RUB": 3.7375,
    "SGD": 0.21828,
    "SKK": 3.7276,
    "THB": 4.8793,
    "TRY": 0.2008,
    "USD": 0.14309,
    "ZAR": 1.0261,
    "EUR": 0.10815
  }
}