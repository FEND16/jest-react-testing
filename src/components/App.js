import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { loadRates, mapObjectToArray } from '../api';
import Button from './Button';

export default class App extends Component {
  state = {
    rates: this.props.rates,
    date: this.props.date,
    base: this.props.base,
    error: '',
    search: ''
  };

  componentDidMount() {
    this.updateRates();
  }

  onChange = ({ target }) => this.setState({ [target.name] : target.value });

  updateRates = () => {
    /* 
     * function for getting data and formatting data is
     * located in `api/index.js`
     */
    loadRates(this.state.date, this.state.base)
      .then(({ rates, date, base }) => {
        this.setState({ rates: mapObjectToArray(rates) , date, base })
      })
      .catch(e => this.setState({ error: e.message }));
  };

  filterList = rate => {
    if(rate.value && this.state.search) {
      return rate.key.includes(this.state.search)
    }
    return rate;
  }

  renderList = rates => rates
    .filter(this.filterList)
    .map(rate => (
      <p className="py-2 px-4 mb-4 border rounded shadow" key={rate.key}>
        {rate.value} {rate.key}
      </p>
   ));

  render() {
    return (
      <div className="flex flex-col justify-center items-center w-1/2 mx-auto font-mono text-grey-darker mt-6">
        <div className="fixed pin-t pin-r m-4 font-mono uppercase text-sm flex flex-col items-center">
          <p>Updated: {this.state.date}</p>
          <Button onClick={this.updateRates} />
        </div>
        <h1 className="text-pink-dark mb-4">{this.state.base} Rates</h1>
        <p className="error">{this.state.error}</p>
        <input 
          type="text" 
          value={this.state.search} 
          onChange={this.onChange} 
          name="search" 
          className="shadow appearance-none border rounded w-full py-2 px-3 mb-4 text-grey-darker" 
          data-test="input"
        />
        <div data-test="list">
          {this.renderList(this.state.rates)}
        </div>
      </div>
    );
  }
}

App.propTypes = {
  rates: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    value: PropTypes.number
  })),
  base: PropTypes.string,
  date: PropTypes.string
};

/* 
 * If no props are sent down from `index.js` the component
 * uses these props, default is EUR and todays rate
 */
App.defaultProps = {
  rates: [],
  base: 'EUR',
  date: 'latest'
};
