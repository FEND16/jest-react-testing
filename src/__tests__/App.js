import React from 'react';
import { shallow, mount, render } from 'enzyme';
import toJSON from 'enzyme-to-json';
import fetchMock from 'fetch-mock';
import App from '../components/App';
import responseObject from '../responseObject';
import responseObjectSEK from '../responseObjectSEK'
import { mapObjectToArray, loadRates } from '../api';

/**
 * Helper function to force all the promises to resolve.
 * Needed when we must wait for a component to load
 * data on mount. Lifecycle methods will only run
 * when we use mount, not when we use shallow or render
 */
function flushAllPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

/**
 * Restore fetchMock after each test. Cleanup duty.
 */
afterEach(() => {
  fetchMock.restore();
  fetchMock.reset();
});

/**
 * I am using a mocked copy of the response object instead of calling
 * the API. These test are only checking if the components renders as it
 * should, not checking if the API is being called. For this purpose
 * we do not need to call the API because that would require much more work
 */

it('populate list with rates', () => {
  /* Mounts the whole app */
  const rates = mapObjectToArray(responseObject.rates);
  const wrapper = shallow(<App rates={rates} />);
  const list = wrapper.find('[data-test="list"]');
  expect(list.children()).toHaveLength(31);
});

it('first rate should be AUD', () => {
  const rates = mapObjectToArray(responseObject.rates);
  const wrapper = shallow(<App rates={rates} />);
  const list = wrapper.find('[data-test="list"]');
  const firstParagraph = list.find('p').first();
  expect(firstParagraph.text()).toContain('AUD');
});

it('should be base EUR', () => {
  const wrapper = shallow(<App />);
  expect(wrapper.find('h1').text()).toContain('EUR');
});

it('should match snapshot', () => {
  const rates = mapObjectToArray(responseObject.rates);
  const wrapper = render(<App rates={rates} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

/**
 * SOLUTIONS TO EXERCISES
 */

/* Just don't pass anything to the App and check if list is
 * shorter 
 */
it('list should not be populated', () => {
  const wrapper = shallow(<App />);
  expect(wrapper.find('[data-test="list"]').children()).toHaveLength(0);
});

/**
 * Fake the error message, pass a failing url to the API-function
 * and expect the message to be shown
 */
it('error message displays when fetch fails', () => {
  const message = 'You failed yo!';
  fetchMock.get('https://api.fixer.io/latest?base=EUR', responseObject);
  fetchMock.get('https://api.fixer.io/fail?base=fail', { throws: { message } });
  const wrapper = mount(<App base="fail" date="fail" />);
  return flushAllPromises().then(() => {
    expect(wrapper.find('.error').text()).toEqual(message);
  });
});

/**
 * Easier to test if the message is being set in the state
 * set state and see if it is visible
 */
it('error message displays when in state', () => {
  const message = 'You failed yo!';
  const wrapper = shallow(<App />);
  wrapper.setState({ error: message });
  expect(wrapper.find('.error').text()).toEqual(message);
});

/**
 * You can either call the api or fake passing the date. If you want to
 * randomize it you can make it pass either yesterday or today by writing
 * a random function. We can check multiple scenarios inside the expect()
 */
it('should check if rates have been updated today or yesterday', () => {
  /* Helper function to left pad a date, so '9' becomes '09' */
  const pad = str => (str.length === 1 ? `0${str}` : str);
  /* Weird stuff to get today and yesterday */
  const date = new Date();
  const yearAndMonth = date.toISOString().slice(0, 8);
  const yesterday = yearAndMonth + date.getDate();
  const today = yearAndMonth + pad((date.getDate() - 1).toString());
  const wrapper = shallow(<App date={today} />);
  const updated = wrapper.find('.fixed p').text();
  expect((updated.includes(today) || updated.includes(yesterday))).toBeTruthy();
});

/**
 * API TESTING, put it in a describe block to exclude all tests with skip
 */

describe.skip('API-testing', () => {
  /* loadRates returns an object, if we turn it into a string
   * and then parse it to json again we can check if the object
   * is valid JSON. If it is not valid JSON, this action should throw
   * an error */
  it('data from api should be json', () => {
    return loadRates().then(returnData => {
      expect(() => {
        JSON.parse(JSON.stringify(returnData));
      }).not.toThrow();
    });
  });
  /**
   * The function does not throw but returns an error message when
   * called with wrong url, check for that error instead.
   */
  it('should return error when called with wrong url', () => {
    return loadRates('fail', 'fail').then(returnData => {
      expect(returnData.error).toBe('Not found');
    });
  });
});

describe('mapObjectToArray', () => {
  it('mapObjectToArray should return array', () => {
    const testObject = { AUD: 1.5, SEK: 1.6 };
    const result = mapObjectToArray(testObject);
    expect(Array.isArray(result)).toBeTruthy();
  });

  /**
   * Sorry about this one, you have to put in an if into
   * the actual function to make it throw. I did that, check in
   * `api/index.js`
   */
  it('mapObjectToArray should throw an error', () => {
    const testObject = 'hello';
    expect(() => {
      mapObjectToArray(testObject);
    }).toThrowError('Not an object');
  });
});

describe('Base and Date', () => {
  it('should render SEK', async () => {
    const wrapper = shallow(<App base="SEK" />);
    expect(wrapper.find('h1').text()).toContain('SEK');
  });

  it('should render different date', () => {
    const date = '2007-02-27';
    const wrapper = shallow(<App date={date} />);
    expect(wrapper.find('.fixed p').text()).toContain(date);
  });

  it('should fetch from api at given date and currency', () => {
    const date = '2007-02-27';
    const base = 'SEK'
    /* mock the fetch but send along a different response, I created a different
     * response object and changed the url 
     */
    fetchMock.get(`https://api.fixer.io/${date}?base=${base}`, responseObjectSEK);
    const wrapper = mount(<App base={base} date={date} />);
    return flushAllPromises()
      .then(() => {
        expect(wrapper.find('h1').text()).toContain(base);
        expect(wrapper.find('.fixed p').text()).toContain(date);
    });
  });
});

describe('Input field test', () => {
  it('should search for currency', () => {
    const search = 'BGN';
    const rates = mapObjectToArray(responseObject.rates);
    const wrapper = shallow(<App rates={rates}/>);
    wrapper.setState({ search: search });
    expect(wrapper.find(`[data-test="list"] p`).first().text()).toContain(search)
  });

  it('list should be empty when search is wrong', () => {
    const search = 'blaj';
    const rates = mapObjectToArray(responseObject.rates);
    const wrapper = shallow(<App rates={rates}/>);
    /* Change the state to simulate a search instead of simulating 
     * an onchange, see below */
    wrapper.setState({ search: search });
    expect(wrapper.find(`[data-test="list"]`).children()).toHaveLength(0);
  });

  /**
   * This one is a bit akward. The docs and API has changed a bit from
   * 2 -> 3 which makes it hard to find a good way to change the node 
   * value. It is easier to just change the state instead of doing this
   */
  it('should update search field', () => {
    //Mocking because of mount triggers componentDidMount()
    fetchMock.get('https://api.fixer.io/latest?base=EUR', responseObject);
    const rates = mapObjectToArray(responseObject.rates);
    const wrapper = mount(<App rates={rates}/>);
    expect(wrapper.find("input").instance().value).toBe('');
    //wrapper.setState({ search: 'AUD' })
    wrapper.find("input").instance().value = 'AUD';
    wrapper.find('input').simulate('change');
    expect(wrapper.find("input").instance().value).toBe('AUD');
  })
});