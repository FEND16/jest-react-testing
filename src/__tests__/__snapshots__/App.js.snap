// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  class="flex flex-col justify-center items-center w-1/2 mx-auto font-mono text-grey-darker mt-6"
>
  <div
    class="fixed pin-t pin-r m-4 font-mono uppercase text-sm flex flex-col items-center"
  >
    <p>
      Updated: latest
    </p>
    <button
      class="bg-pink hover:bg-pink-light 
      text-white 
      font-bold py-2 px-4 
      border-b-4 
      border-pink-dark 
      hover:border-pink rounded mt-4
    "
      data-test="button"
    >
      Update Rates
    </button>
  </div>
  <h1
    class="text-pink-dark mb-4"
  >
    EUR Rates
  </h1>
  <p
    class="error"
  >
    
  </p>
  <input
    class="shadow appearance-none border rounded w-full py-2 px-3 mb-4 text-grey-darker"
    date-test="input"
    name="search"
    type="text"
    value=""
  >
    
  </input>
  <div
    data-test="list"
  >
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1.5316 AUD
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1.9558 BGN
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      3.9196 BRL
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1.5028 CAD
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1.1772 CHF
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      7.7975 CNY
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      25.84 CZK
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      7.445 DKK
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      0.88593 GBP
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      9.2953 HKD
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      7.5398 HRK
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      310.99 HUF
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      16123 IDR
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      4.1432 ILS
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      76.312 INR
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      134.7 JPY
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1277.7 KRW
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      23.522 MXN
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      4.8481 MYR
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      9.8605 NOK
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1.6825 NZD
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      59.43 PHP
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      4.1884 PLN
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      4.6499 RON
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      68.597 RUB
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      9.8727 SEK
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1.5937 SGD
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      39.027 THB
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      4.5475 TRY
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      1.1895 USD
    </p>
    <p
      class="py-2 px-4 mb-4 border rounded shadow"
    >
      14.811 ZAR
    </p>
  </div>
</div>
`;
